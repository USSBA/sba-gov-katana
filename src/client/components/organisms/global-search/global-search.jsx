import React from 'react'
import { connect } from 'react-redux'
import { assign, camelCase, chain, includes, pickBy, startCase } from 'lodash'
import { bindActionCreators } from 'redux'

import styles from './global-search.scss'
import { ApplyButton, MultiSelect } from 'atoms'
import * as ContentActions from '../../../actions/content.js'
import { logPageEvent } from '../../../services/analytics.js'
import { getQueryParams } from '../../../services/utils.js'
import { logEvent } from '../../../services/analytics.js'

const createSlug = str => {
  return str
    .toLowerCase()
    .replace(/[^\w\s-]/g, '') // remove non-word [a-z0-9_], non-whitespace, non-hyphen characters
    .replace(/[\s_-]+/g, '-') // swap any length of whitespace, underscore, hyphen characters with a single -
    .replace(/^-+|-+$/g, '')
}

export class GlobalSearch extends React.PureComponent {
  constructor(ownProps) {
    super()
    this.state = {
      query: {}
    }
  }

  componentWillMount() {
    this.props.actions.fetchContentIfNeeded('taxonomies', 'taxonomys', {
      names: this.props.taxonomyFilters.join(',')
    })
  }

  renderMultiSelects(taxonomies) {
    const _multiselects = taxonomies.map(taxonomy => {
      const { name } = taxonomy
      let newName
      const id = `${createSlug(name)}-select`
      const stateName = camelCase(name)
      const includesAllInTaxonomy = ['All', ...taxonomy.terms]
      const options = includesAllInTaxonomy.map(entry => {
        return { label: entry, value: entry }
      })

      name === 'businessStage' ? (newName = 'courseTopic') : newName

      const _ms = {
        id: id,
        onChange: event => {
          this.handleChange(event, stateName)
        },
        name: id,
        label: startCase(newName || name),
        value: this.state.query[name] || 'All',
        options: options
      }

      return _ms
    })

    return _multiselects.map((multiSelectProps, index) => {
      const returnNull = () => {
        return null
      }

      return (
        <div className={styles.multiSelect} key={index}>
          <MultiSelect
            {...multiSelectProps}
            onBlur={returnNull}
            onFocus={returnNull}
            validationState=""
            errorText=""
            autoFocus={false}
            multi={false}
          />
        </div>
      )
    })
  }

  handleChange(event, selectStateKey) {
    this.handleQueryChange(selectStateKey, event.value)
  }

  handleQueryChange(field, value) {
    if (field !== 'searchTerm') {
      // Log Analytic Event, but not for search term
      this.fireDocumentationLookupEvent(`${field}: ${value}`)
    }
    const newQueryFieldValue = {}
    newQueryFieldValue[field] = value
    const currentQuery = this.state.query
    const newQuery = _.assign({}, currentQuery, newQueryFieldValue)
    console.log('CHANGES', newQuery)
    this.setState({ query: newQuery })
  }

  fireDocumentationLookupEvent(action, value = null) {
    this.fireEvent('course-topic-lookup', action, value)
  }

  fireEvent(category, action, value) {
    logEvent({
      category: category,
      action: action,
      label: window.location.pathname,
      value: value
    })
  }

  render() {
    return (
      <div>
        <div className={styles.banner}>
          <h2 className={styles.header}>{this.props.title}</h2>
          {this.props.taxonomies && (
            <div>
              {this.renderMultiSelects(this.props.taxonomies)}
              <div className={styles.applyButton}>
                <ApplyButton submit={this.props.onSubmit} />
              </div>
            </div>
          )}
        </div>
        <div />
      </div>
    )
  }
}

function mapReduxStateToProps(reduxState) {
  return {
    taxonomies: reduxState.contentReducer.taxonomies
  }
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators(ContentActions, dispatch)
  }
}

export default connect(mapReduxStateToProps, mapDispatchToProps)(GlobalSearch)
