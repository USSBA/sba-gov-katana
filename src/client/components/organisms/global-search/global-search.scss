@include theme(sba-blue, '.banner', $sba-blue, #fff, #fff);

.banner {
  position: relative;

  width: 100%;

  @include grid-media($sba-grid-small) {
    padding: 20px;
  }
  @include grid-media($sba-grid-medium) {
    padding: 20px 30px 20px 30px;
  }
  @include grid-media($sba-grid-large) {
    height: 220px;
    padding: 40px 28px 20px 28px;

    background-image: url('../../../../../public/assets/images/tensteps/title-section-pattern.png');
    background-repeat: no-repeat;
    background-position: right;
    background-size: 40%;
  }
}

.header {
  margin: 0 0 20px 0;
}

.multiSelect {
  @include grid-media($sba-grid-small) {
    width: 100%;
  }
  @include grid-media($sba-grid-large) {
    display: inline-block;

    width: 140px;
    margin-right: 20px;
    &:last-child {
      margin-right: 0;
    }
  }
}

@media screen and (min-width: 1200px) {
  .multiSelect {
    width: 13%;
  }
}

.applyButton {
  display: inline-block;
  margin-top: 33px;
  vertical-align: top;
}

.table {
  border-bottom: 1px solid #dadada;

  @include grid-media($sba-grid-medium-and-small) {
    display: block;
    overflow-y: auto;

    height: 600px;
    & thead {
      display: none;
    }
    & tbody,
    tr,
    td {
      display: block;
    }
    & tr {
      border-bottom: 1px solid #dadada;
    }
    & td {
      overflow: auto;

      border: none;
    }
    & .mobileHeader {
      font-weight: bold;

      float: left;

      width: 23%;
      margin-left: 2%;
      margin-right: 5%;
    }
  }
  @include grid-media($sba-grid-large) {
    & .mobileHeader {
      display: none;
    }
    // & .nameAndAddressCol {
    //   width: 27vw !important;
    // }
    // & .nameAndAddressHead {
    //   width: 26.4vw !important;
    // }
    // & .activeSinceCol {
    //   width: 10vw !important;
    // }
    // & .activeSinceHead {
    //   width: 9.9vw !important;
    // }
    // & .contactInfoHead {
    //   width: 10vw !important;
    // }
    & tbody,
    thead {
      display: block;
    }
    & tbody {
      overflow-x: hidden;
      overflow-y: auto;

      height: 600px;
    }
    & tbody td {
      width: 21% !important;
      padding: 10px;

      border: none;
    }
    & thead {
      display: block;

      width: 100%;
    }
    & thead tr {
      display: block;

      width: 100%;
    }
    & thead th {
      display: inline-block;

      width: 20.5vw !important;
      padding: 10px;

      border: none;
    }
    & tr {
      border-bottom: 1px solid #dadada;
    }
  }
}
