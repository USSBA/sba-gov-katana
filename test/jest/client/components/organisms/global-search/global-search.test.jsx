import React from 'react'
import renderer from 'react-test-renderer'
import _ from 'lodash'
import { shallow } from 'enzyme'

import { GlobalSearch } from 'organisms/global-search/global-search'

jest.mock('../../../../../../src/client/services/utils.js')

const fetchContentIfNeeded = jest.fn()

const testPropsOne = {
  title: 'My Courses',
  type: 'courses',
  taxonomyFilters: ['businessStage'],
  actions: {
    fetchContentIfNeeded: fetchContentIfNeeded
  },
  location: {
    query: {
      topic: 'All'
    }
  }
}

const testPropsTwo = {
  title: 'My Courses',
  type: 'courses',
  taxonomyFilters: ['businessStage', 'program'],
  actions: {
    fetchContentIfNeeded: fetchContentIfNeeded
  },
  location: {
    query: {
      topic: 'Starting a business'
    }
  }
}

describe('GlobalSearch', () => {
  test('should render with one taxonomyFilter', () => {
    const props = _.clone(testPropsOne)
    const component = renderer.create(<GlobalSearch {...props} />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('should render with one taxonomyFilter', () => {
    const props = _.clone(testPropsOne)
    const component = renderer.create(<GlobalSearch {...props} />)
    const tree = component.toJSON()
  })

  test('should render with two taxonomyFilter', () => {
    const props = _.clone(testPropsTwo)
    const component = renderer.create(<GlobalSearch {...props} />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
