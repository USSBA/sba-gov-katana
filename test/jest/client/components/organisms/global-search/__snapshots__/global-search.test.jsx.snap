// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GlobalSearch should render with one taxonomyFilter 1`] = `
<div>
  <div
    className="banner"
  >
    <h2
      className="header"
    >
      My Courses
    </h2>
    <form
      onSubmit={[Function]}
    >
      <div
        className="applyButton"
      >
        <button
          className="button alternate large primary responsive spacing"
          data-testid="button"
          type="submit"
        >
          Apply
        </button>
      </div>
    </form>
  </div>
</div>
`;

exports[`GlobalSearch should render with two taxonomyFilter 1`] = `
<div>
  <div
    className="banner"
  >
    <h2
      className="header"
    >
      My Courses
    </h2>
    <form
      onSubmit={[Function]}
    >
      <div
        className="applyButton"
      >
        <button
          className="button alternate large primary responsive spacing"
          data-testid="button"
          type="submit"
        >
          Apply
        </button>
      </div>
    </form>
  </div>
</div>
`;
